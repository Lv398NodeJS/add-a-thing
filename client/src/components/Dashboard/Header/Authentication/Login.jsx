import React, { Component } from 'react';
import {
  Form, Button, ButtonGroup, ToggleButton, Container, Row, Col, Alert,
} from 'react-bootstrap';
import { Redirect } from 'react-router-dom';
import { Toaster, Intent } from '@blueprintjs/core';
import db, { facebookProvider, googleProvider } from '@src/fire';
import connect from 'react-redux/es/connect/connect';
import NavBar from '../Header';
import * as authAction from '../../../../actions/authAction';

class Login extends Component {
  loginWithEmailPassword = (event) => {
    event.preventDefault();
    const email = this.emailInput.value;
    const password = this.passwordInput.value;
    const { authWithEmailPassword } = this.props;
    authWithEmailPassword({ email, password });
  }

  render() {
    const { redirect, user } = this.props;
    console.log('Do redirect', redirect, user);
    if (user) {
      return <Redirect to="/" />;
    }
    return (
      <>
        <NavBar isLoggedIn={redirect} />
        <Container className="App">
          <Row>
            <Col md={{ span: 4, offset: 4 }}>
              <Form
                onSubmit={this.loginWithEmailPassword}
                ref={(form) => {
                  this.loginForm = form;
                }}
              >
                <h3>Log in to Add a Thing</h3>
                <Form.Text className="text-muted">
                  <Alert variant="secondary">
                    <Alert.Heading>NOTE</Alert.Heading>
                    <p>
                      If you do not have an account yet, this form will create an account for you.
                    </p>
                  </Alert>
                </Form.Text>
                <Form.Group controlId="formBasicEmail">
                  <Form.Control
                    type="email"
                    placeholder="email or username"
                    ref={(input) => {
                      this.emailInput = input;
                    }}
                  />
                </Form.Group>

                <Form.Group controlId="formBasicPassword">
                  <Form.Control
                    type="password"
                    placeholder="password"
                    ref={(input) => {
                      this.passwordInput = input;
                    }}
                  />
                </Form.Group>
                <Button variant="primary" size="md" block type="submit" valur="Log In">Log in</Button>
              </Form>

              <div className="d-flex flex-column">
                <ButtonGroup toggle className="mt-2">
                  <ToggleButton
                    type="radio"
                    name="radio"
                    defaultChecked
                    value="1"
                    onClick={() => {
                      this.authWithGoogle();
                    }}
                  >

                    Log in with Google
                  </ToggleButton>
                  <ToggleButton
                    type="radio"
                    name="radio"
                    value="2"
                    onClick={() => {
                      this.authWithFacebook();
                    }}
                  >

                    Log in with Facebook
                  </ToggleButton>
                </ButtonGroup>
              </div>
            </Col>
          </Row>
        </Container>
      </>
    );
  }
}
