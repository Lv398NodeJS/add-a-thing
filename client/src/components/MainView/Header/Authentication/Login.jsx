import React, { Component } from 'react';
import {
  Form, Button, Container, Row, Col, Alert,
} from 'react-bootstrap';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { Redirect } from 'react-router-dom';
import * as loginActions from '@actions/loginationActions';
import Header from '../Header';


export class Login extends Component {
  constructor(props) {
    super(props);
    this.state = {
      redirect: false,
      showAlertLogin: false,
    };
  }

  logination = async () => {
    const { loginationActions: { loginUser } } = this.props;
    const { email, password } = this.state;
    // eslint-disable-next-line no-restricted-globals
    event.preventDefault();
    const loginUserData = {
      email: email,
      password: password,
    };
    if(email && password) {
        loginUser(loginUserData);
    } else {this.setState({showAlertLogin: true});}
  };

  componentWillReceiveProps(userData) {
    if (userData) {
      this.setState({redirect: true});
    }
  };

  handleSave = (event) => {
    const { name, value } = event.target;
    this.setState({
      [name]: value,
    });
  };

  render() {
    const { redirect,showAlertLogin, email, password } = this.state;
    if (redirect === true) {
      return <Redirect to="/" />;
    }

    const alertLogin = showAlertLogin ? (<Alert variant='danger'>
      Wrong email or password
    </Alert>): <></>;

    return (
      <>
        <Header />
        <Container className="App">
          <Row>
            <Col md={{ span: 4, offset: 4 }}>
              <h3>Log in to Add a Thing</h3>
              <Form.Text className="text-muted">
                <Alert variant="secondary">
                  <Alert.Heading>NOTE</Alert.Heading>
                  <p>
                      If you do not have an account yet, please click on Signup.
                  </p>
                </Alert>
              </Form.Text>
              <Form.Group controlId="formBasicEmail">
                <Form.Control
                  type="email"
                  placeholder="email or username"
                  name="email"
                  value={email}
                  onChange={this.handleSave}
                />
              </Form.Group>

              <Form.Group controlId="formBasicPassword">
                <Form.Control
                  type="password"
                  placeholder="password"
                  name="password"
                  value={password}
                  onChange={this.handleSave}
                />
              </Form.Group>
              <>{alertLogin}</>
              <Button
                variant="primary"
                size="md"
                type="submit"
                valur="Create New Account"
                onClick={this.logination}
              >
                  Log in
              </Button>
            </Col>
          </Row>
        </Container>
      </>
    );
  }
}

const mapStateToProps = state => ({
  token: state.loginationReducer.token,
  userData: state.loginationReducer.userData,
  loginedData: state.loginationReducer.loginedData,
});

const mapDispatchToProps = dispatch => ({
  loginationActions: bindActionCreators(loginActions, dispatch),
});

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(Login);
