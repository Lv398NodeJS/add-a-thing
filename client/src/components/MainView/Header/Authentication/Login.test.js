import React from 'react';
import { shallow } from 'enzyme';
import { Login } from './Login';

describe('Login', () => {
  it('Login should render correctly', () => {
    const component = shallow(<Login />);
    expect(component).toMatchSnapshot();
  });
});
