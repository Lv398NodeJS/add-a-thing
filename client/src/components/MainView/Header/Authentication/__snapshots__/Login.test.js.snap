// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Login should render correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Login />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <UNDEFINED />,
        <ForwardRef
          className="App"
          fluid={false}
        >
          <ForwardRef(render)>
            <ForwardRef
              md={
                Object {
                  "offset": 4,
                  "span": 4,
                }
              }
            >
              <h3>
                Log in to Add a Thing
              </h3>
              <ForwardRef
                className="text-muted"
              >
                <ForwardRef
                  closeLabel="Close alert"
                  show={true}
                  transition={[Function]}
                  variant="secondary"
                >
                  <ForwardRef>
                    NOTE
                  </ForwardRef>
                  <p>
                    If you do not have an account yet, please click on Signup.
                  </p>
                </ForwardRef>
              </ForwardRef>
              <ForwardRef
                controlId="formBasicEmail"
              >
                <ForwardRef
                  placeholder="email or username"
                  type="email"
                />
              </ForwardRef>
              <ForwardRef
                controlId="formBasicPassword"
              >
                <ForwardRef
                  placeholder="password"
                  type="password"
                />
              </ForwardRef>
              <React.Fragment>
                <React.Fragment />
              </React.Fragment>
              <ForwardRef
                active={false}
                disabled={false}
                onClick={[Function]}
                size="md"
                type="submit"
                valur="Create New Account"
                variant="primary"
              >
                Log in
              </ForwardRef>
            </ForwardRef>
          </ForwardRef(render)>
        </ForwardRef>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": Object {
          "$$typeof": Symbol(react.memo),
          "WrappedComponent": [Function],
          "compare": null,
          "displayName": "Connect(Header)",
          "type": [Function],
          Symbol(Symbol.hasInstance): [Function],
        },
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": <ForwardRef(render)>
            <ForwardRef
              md={
                Object {
                  "offset": 4,
                  "span": 4,
                }
              }
            >
              <h3>
                Log in to Add a Thing
              </h3>
              <ForwardRef
                className="text-muted"
              >
                <ForwardRef
                  closeLabel="Close alert"
                  show={true}
                  transition={[Function]}
                  variant="secondary"
                >
                  <ForwardRef>
                    NOTE
                  </ForwardRef>
                  <p>
                    If you do not have an account yet, please click on Signup.
                  </p>
                </ForwardRef>
              </ForwardRef>
              <ForwardRef
                controlId="formBasicEmail"
              >
                <ForwardRef
                  placeholder="email or username"
                  type="email"
                />
              </ForwardRef>
              <ForwardRef
                controlId="formBasicPassword"
              >
                <ForwardRef
                  placeholder="password"
                  type="password"
                />
              </ForwardRef>
              <React.Fragment>
                <React.Fragment />
              </React.Fragment>
              <ForwardRef
                active={false}
                disabled={false}
                onClick={[Function]}
                size="md"
                type="submit"
                valur="Create New Account"
                variant="primary"
              >
                Log in
              </ForwardRef>
            </ForwardRef>
          </ForwardRef(render)>,
          "className": "App",
          "fluid": false,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": <ForwardRef
              md={
                Object {
                  "offset": 4,
                  "span": 4,
                }
              }
            >
              <h3>
                Log in to Add a Thing
              </h3>
              <ForwardRef
                className="text-muted"
              >
                <ForwardRef
                  closeLabel="Close alert"
                  show={true}
                  transition={[Function]}
                  variant="secondary"
                >
                  <ForwardRef>
                    NOTE
                  </ForwardRef>
                  <p>
                    If you do not have an account yet, please click on Signup.
                  </p>
                </ForwardRef>
              </ForwardRef>
              <ForwardRef
                controlId="formBasicEmail"
              >
                <ForwardRef
                  placeholder="email or username"
                  type="email"
                />
              </ForwardRef>
              <ForwardRef
                controlId="formBasicPassword"
              >
                <ForwardRef
                  placeholder="password"
                  type="password"
                />
              </ForwardRef>
              <React.Fragment>
                <React.Fragment />
              </React.Fragment>
              <ForwardRef
                active={false}
                disabled={false}
                onClick={[Function]}
                size="md"
                type="submit"
                valur="Create New Account"
                variant="primary"
              >
                Log in
              </ForwardRef>
            </ForwardRef>,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": Array [
                <h3>
                  Log in to Add a Thing
                </h3>,
                <ForwardRef
                  className="text-muted"
                >
                  <ForwardRef
                    closeLabel="Close alert"
                    show={true}
                    transition={[Function]}
                    variant="secondary"
                  >
                    <ForwardRef>
                      NOTE
                    </ForwardRef>
                    <p>
                      If you do not have an account yet, please click on Signup.
                    </p>
                  </ForwardRef>
                </ForwardRef>,
                <ForwardRef
                  controlId="formBasicEmail"
                >
                  <ForwardRef
                    placeholder="email or username"
                    type="email"
                  />
                </ForwardRef>,
                <ForwardRef
                  controlId="formBasicPassword"
                >
                  <ForwardRef
                    placeholder="password"
                    type="password"
                  />
                </ForwardRef>,
                <React.Fragment>
                  <React.Fragment />
                </React.Fragment>,
                <ForwardRef
                  active={false}
                  disabled={false}
                  onClick={[Function]}
                  size="md"
                  type="submit"
                  valur="Create New Account"
                  variant="primary"
                >
                  Log in
                </ForwardRef>,
              ],
              "md": Object {
                "offset": 4,
                "span": 4,
              },
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "Log in to Add a Thing",
                },
                "ref": null,
                "rendered": "Log in to Add a Thing",
                "type": "h3",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "children": <ForwardRef
                    closeLabel="Close alert"
                    show={true}
                    transition={[Function]}
                    variant="secondary"
                  >
                    <ForwardRef>
                      NOTE
                    </ForwardRef>
                    <p>
                      If you do not have an account yet, please click on Signup.
                    </p>
                  </ForwardRef>,
                  "className": "text-muted",
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {
                    "children": Array [
                      <ForwardRef>
                        NOTE
                      </ForwardRef>,
                      <p>
                        If you do not have an account yet, please click on Signup.
                      </p>,
                    ],
                    "closeLabel": "Close alert",
                    "show": true,
                    "transition": [Function],
                    "variant": "secondary",
                  },
                  "ref": null,
                  "rendered": Array [
                    Object {
                      "instance": null,
                      "key": undefined,
                      "nodeType": "function",
                      "props": Object {
                        "children": "NOTE",
                      },
                      "ref": null,
                      "rendered": "NOTE",
                      "type": Object {
                        "$$typeof": Symbol(react.forward_ref),
                        "defaultProps": undefined,
                        "displayName": "AlertHeading",
                        "render": [Function],
                      },
                    },
                    Object {
                      "instance": null,
                      "key": undefined,
                      "nodeType": "host",
                      "props": Object {
                        "children": "If you do not have an account yet, please click on Signup.",
                      },
                      "ref": null,
                      "rendered": "If you do not have an account yet, please click on Signup.",
                      "type": "p",
                    },
                  ],
                  "type": Object {
                    "$$typeof": Symbol(react.forward_ref),
                    "Heading": Object {
                      "$$typeof": Symbol(react.forward_ref),
                      "defaultProps": undefined,
                      "displayName": "AlertHeading",
                      "render": [Function],
                    },
                    "Link": Object {
                      "$$typeof": Symbol(react.forward_ref),
                      "defaultProps": undefined,
                      "displayName": "AlertLink",
                      "render": [Function],
                    },
                    "defaultProps": Object {
                      "closeLabel": "Close alert",
                      "show": true,
                      "transition": [Function],
                    },
                    "displayName": "Alert",
                    "render": [Function],
                  },
                },
                "type": Object {
                  "$$typeof": Symbol(react.forward_ref),
                  "displayName": "FormText",
                  "render": [Function],
                },
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "children": <ForwardRef
                    placeholder="email or username"
                    type="email"
                  />,
                  "controlId": "formBasicEmail",
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {
                    "placeholder": "email or username",
                    "type": "email",
                  },
                  "ref": [Function],
                  "rendered": null,
                  "type": Object {
                    "$$typeof": Symbol(react.forward_ref),
                    "Feedback": Object {
                      "$$typeof": Symbol(react.forward_ref),
                      "defaultProps": Object {
                        "type": "valid",
                      },
                      "displayName": "Feedback",
                      "propTypes": Object {
                        "as": [Function],
                        "type": [Function],
                      },
                      "render": [Function],
                    },
                    "displayName": "FormControl",
                    "render": [Function],
                  },
                },
                "type": Object {
                  "$$typeof": Symbol(react.forward_ref),
                  "displayName": "FormGroup",
                  "render": [Function],
                },
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "children": <ForwardRef
                    placeholder="password"
                    type="password"
                  />,
                  "controlId": "formBasicPassword",
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {
                    "placeholder": "password",
                    "type": "password",
                  },
                  "ref": [Function],
                  "rendered": null,
                  "type": Object {
                    "$$typeof": Symbol(react.forward_ref),
                    "Feedback": Object {
                      "$$typeof": Symbol(react.forward_ref),
                      "defaultProps": Object {
                        "type": "valid",
                      },
                      "displayName": "Feedback",
                      "propTypes": Object {
                        "as": [Function],
                        "type": [Function],
                      },
                      "render": [Function],
                    },
                    "displayName": "FormControl",
                    "render": [Function],
                  },
                },
                "type": Object {
                  "$$typeof": Symbol(react.forward_ref),
                  "displayName": "FormGroup",
                  "render": [Function],
                },
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "children": <React.Fragment />,
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {},
                  "ref": null,
                  "rendered": null,
                  "type": Symbol(react.fragment),
                },
                "type": Symbol(react.fragment),
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "active": false,
                  "children": "Log in",
                  "disabled": false,
                  "onClick": [Function],
                  "size": "md",
                  "type": "submit",
                  "valur": "Create New Account",
                  "variant": "primary",
                },
                "ref": null,
                "rendered": "Log in",
                "type": Object {
                  "$$typeof": Symbol(react.forward_ref),
                  "defaultProps": Object {
                    "active": false,
                    "disabled": false,
                    "type": "button",
                    "variant": "primary",
                  },
                  "displayName": "Button",
                  "render": [Function],
                },
              },
            ],
            "type": Object {
              "$$typeof": Symbol(react.forward_ref),
              "displayName": "Col",
              "render": [Function],
            },
          },
          "type": Object {
            "$$typeof": Symbol(react.forward_ref),
            "defaultProps": undefined,
            "displayName": "Bootstrap(Row)",
            "propTypes": undefined,
            "render": [Function],
          },
        },
        "type": Object {
          "$$typeof": Symbol(react.forward_ref),
          "defaultProps": Object {
            "fluid": false,
          },
          "displayName": "Container",
          "render": [Function],
        },
      },
    ],
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <UNDEFINED />,
          <ForwardRef
            className="App"
            fluid={false}
          >
            <ForwardRef(render)>
              <ForwardRef
                md={
                  Object {
                    "offset": 4,
                    "span": 4,
                  }
                }
              >
                <h3>
                  Log in to Add a Thing
                </h3>
                <ForwardRef
                  className="text-muted"
                >
                  <ForwardRef
                    closeLabel="Close alert"
                    show={true}
                    transition={[Function]}
                    variant="secondary"
                  >
                    <ForwardRef>
                      NOTE
                    </ForwardRef>
                    <p>
                      If you do not have an account yet, please click on Signup.
                    </p>
                  </ForwardRef>
                </ForwardRef>
                <ForwardRef
                  controlId="formBasicEmail"
                >
                  <ForwardRef
                    placeholder="email or username"
                    type="email"
                  />
                </ForwardRef>
                <ForwardRef
                  controlId="formBasicPassword"
                >
                  <ForwardRef
                    placeholder="password"
                    type="password"
                  />
                </ForwardRef>
                <React.Fragment>
                  <React.Fragment />
                </React.Fragment>
                <ForwardRef
                  active={false}
                  disabled={false}
                  onClick={[Function]}
                  size="md"
                  type="submit"
                  valur="Create New Account"
                  variant="primary"
                >
                  Log in
                </ForwardRef>
              </ForwardRef>
            </ForwardRef(render)>
          </ForwardRef>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": Object {
            "$$typeof": Symbol(react.memo),
            "WrappedComponent": [Function],
            "compare": null,
            "displayName": "Connect(Header)",
            "type": [Function],
            Symbol(Symbol.hasInstance): [Function],
          },
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": <ForwardRef(render)>
              <ForwardRef
                md={
                  Object {
                    "offset": 4,
                    "span": 4,
                  }
                }
              >
                <h3>
                  Log in to Add a Thing
                </h3>
                <ForwardRef
                  className="text-muted"
                >
                  <ForwardRef
                    closeLabel="Close alert"
                    show={true}
                    transition={[Function]}
                    variant="secondary"
                  >
                    <ForwardRef>
                      NOTE
                    </ForwardRef>
                    <p>
                      If you do not have an account yet, please click on Signup.
                    </p>
                  </ForwardRef>
                </ForwardRef>
                <ForwardRef
                  controlId="formBasicEmail"
                >
                  <ForwardRef
                    placeholder="email or username"
                    type="email"
                  />
                </ForwardRef>
                <ForwardRef
                  controlId="formBasicPassword"
                >
                  <ForwardRef
                    placeholder="password"
                    type="password"
                  />
                </ForwardRef>
                <React.Fragment>
                  <React.Fragment />
                </React.Fragment>
                <ForwardRef
                  active={false}
                  disabled={false}
                  onClick={[Function]}
                  size="md"
                  type="submit"
                  valur="Create New Account"
                  variant="primary"
                >
                  Log in
                </ForwardRef>
              </ForwardRef>
            </ForwardRef(render)>,
            "className": "App",
            "fluid": false,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "children": <ForwardRef
                md={
                  Object {
                    "offset": 4,
                    "span": 4,
                  }
                }
              >
                <h3>
                  Log in to Add a Thing
                </h3>
                <ForwardRef
                  className="text-muted"
                >
                  <ForwardRef
                    closeLabel="Close alert"
                    show={true}
                    transition={[Function]}
                    variant="secondary"
                  >
                    <ForwardRef>
                      NOTE
                    </ForwardRef>
                    <p>
                      If you do not have an account yet, please click on Signup.
                    </p>
                  </ForwardRef>
                </ForwardRef>
                <ForwardRef
                  controlId="formBasicEmail"
                >
                  <ForwardRef
                    placeholder="email or username"
                    type="email"
                  />
                </ForwardRef>
                <ForwardRef
                  controlId="formBasicPassword"
                >
                  <ForwardRef
                    placeholder="password"
                    type="password"
                  />
                </ForwardRef>
                <React.Fragment>
                  <React.Fragment />
                </React.Fragment>
                <ForwardRef
                  active={false}
                  disabled={false}
                  onClick={[Function]}
                  size="md"
                  type="submit"
                  valur="Create New Account"
                  variant="primary"
                >
                  Log in
                </ForwardRef>
              </ForwardRef>,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "children": Array [
                  <h3>
                    Log in to Add a Thing
                  </h3>,
                  <ForwardRef
                    className="text-muted"
                  >
                    <ForwardRef
                      closeLabel="Close alert"
                      show={true}
                      transition={[Function]}
                      variant="secondary"
                    >
                      <ForwardRef>
                        NOTE
                      </ForwardRef>
                      <p>
                        If you do not have an account yet, please click on Signup.
                      </p>
                    </ForwardRef>
                  </ForwardRef>,
                  <ForwardRef
                    controlId="formBasicEmail"
                  >
                    <ForwardRef
                      placeholder="email or username"
                      type="email"
                    />
                  </ForwardRef>,
                  <ForwardRef
                    controlId="formBasicPassword"
                  >
                    <ForwardRef
                      placeholder="password"
                      type="password"
                    />
                  </ForwardRef>,
                  <React.Fragment>
                    <React.Fragment />
                  </React.Fragment>,
                  <ForwardRef
                    active={false}
                    disabled={false}
                    onClick={[Function]}
                    size="md"
                    type="submit"
                    valur="Create New Account"
                    variant="primary"
                  >
                    Log in
                  </ForwardRef>,
                ],
                "md": Object {
                  "offset": 4,
                  "span": 4,
                },
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "children": "Log in to Add a Thing",
                  },
                  "ref": null,
                  "rendered": "Log in to Add a Thing",
                  "type": "h3",
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {
                    "children": <ForwardRef
                      closeLabel="Close alert"
                      show={true}
                      transition={[Function]}
                      variant="secondary"
                    >
                      <ForwardRef>
                        NOTE
                      </ForwardRef>
                      <p>
                        If you do not have an account yet, please click on Signup.
                      </p>
                    </ForwardRef>,
                    "className": "text-muted",
                  },
                  "ref": null,
                  "rendered": Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "function",
                    "props": Object {
                      "children": Array [
                        <ForwardRef>
                          NOTE
                        </ForwardRef>,
                        <p>
                          If you do not have an account yet, please click on Signup.
                        </p>,
                      ],
                      "closeLabel": "Close alert",
                      "show": true,
                      "transition": [Function],
                      "variant": "secondary",
                    },
                    "ref": null,
                    "rendered": Array [
                      Object {
                        "instance": null,
                        "key": undefined,
                        "nodeType": "function",
                        "props": Object {
                          "children": "NOTE",
                        },
                        "ref": null,
                        "rendered": "NOTE",
                        "type": Object {
                          "$$typeof": Symbol(react.forward_ref),
                          "defaultProps": undefined,
                          "displayName": "AlertHeading",
                          "render": [Function],
                        },
                      },
                      Object {
                        "instance": null,
                        "key": undefined,
                        "nodeType": "host",
                        "props": Object {
                          "children": "If you do not have an account yet, please click on Signup.",
                        },
                        "ref": null,
                        "rendered": "If you do not have an account yet, please click on Signup.",
                        "type": "p",
                      },
                    ],
                    "type": Object {
                      "$$typeof": Symbol(react.forward_ref),
                      "Heading": Object {
                        "$$typeof": Symbol(react.forward_ref),
                        "defaultProps": undefined,
                        "displayName": "AlertHeading",
                        "render": [Function],
                      },
                      "Link": Object {
                        "$$typeof": Symbol(react.forward_ref),
                        "defaultProps": undefined,
                        "displayName": "AlertLink",
                        "render": [Function],
                      },
                      "defaultProps": Object {
                        "closeLabel": "Close alert",
                        "show": true,
                        "transition": [Function],
                      },
                      "displayName": "Alert",
                      "render": [Function],
                    },
                  },
                  "type": Object {
                    "$$typeof": Symbol(react.forward_ref),
                    "displayName": "FormText",
                    "render": [Function],
                  },
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {
                    "children": <ForwardRef
                      placeholder="email or username"
                      type="email"
                    />,
                    "controlId": "formBasicEmail",
                  },
                  "ref": null,
                  "rendered": Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "function",
                    "props": Object {
                      "placeholder": "email or username",
                      "type": "email",
                    },
                    "ref": [Function],
                    "rendered": null,
                    "type": Object {
                      "$$typeof": Symbol(react.forward_ref),
                      "Feedback": Object {
                        "$$typeof": Symbol(react.forward_ref),
                        "defaultProps": Object {
                          "type": "valid",
                        },
                        "displayName": "Feedback",
                        "propTypes": Object {
                          "as": [Function],
                          "type": [Function],
                        },
                        "render": [Function],
                      },
                      "displayName": "FormControl",
                      "render": [Function],
                    },
                  },
                  "type": Object {
                    "$$typeof": Symbol(react.forward_ref),
                    "displayName": "FormGroup",
                    "render": [Function],
                  },
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {
                    "children": <ForwardRef
                      placeholder="password"
                      type="password"
                    />,
                    "controlId": "formBasicPassword",
                  },
                  "ref": null,
                  "rendered": Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "function",
                    "props": Object {
                      "placeholder": "password",
                      "type": "password",
                    },
                    "ref": [Function],
                    "rendered": null,
                    "type": Object {
                      "$$typeof": Symbol(react.forward_ref),
                      "Feedback": Object {
                        "$$typeof": Symbol(react.forward_ref),
                        "defaultProps": Object {
                          "type": "valid",
                        },
                        "displayName": "Feedback",
                        "propTypes": Object {
                          "as": [Function],
                          "type": [Function],
                        },
                        "render": [Function],
                      },
                      "displayName": "FormControl",
                      "render": [Function],
                    },
                  },
                  "type": Object {
                    "$$typeof": Symbol(react.forward_ref),
                    "displayName": "FormGroup",
                    "render": [Function],
                  },
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {
                    "children": <React.Fragment />,
                  },
                  "ref": null,
                  "rendered": Object {
                    "instance": null,
                    "key": undefined,
                    "nodeType": "function",
                    "props": Object {},
                    "ref": null,
                    "rendered": null,
                    "type": Symbol(react.fragment),
                  },
                  "type": Symbol(react.fragment),
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {
                    "active": false,
                    "children": "Log in",
                    "disabled": false,
                    "onClick": [Function],
                    "size": "md",
                    "type": "submit",
                    "valur": "Create New Account",
                    "variant": "primary",
                  },
                  "ref": null,
                  "rendered": "Log in",
                  "type": Object {
                    "$$typeof": Symbol(react.forward_ref),
                    "defaultProps": Object {
                      "active": false,
                      "disabled": false,
                      "type": "button",
                      "variant": "primary",
                    },
                    "displayName": "Button",
                    "render": [Function],
                  },
                },
              ],
              "type": Object {
                "$$typeof": Symbol(react.forward_ref),
                "displayName": "Col",
                "render": [Function],
              },
            },
            "type": Object {
              "$$typeof": Symbol(react.forward_ref),
              "defaultProps": undefined,
              "displayName": "Bootstrap(Row)",
              "propTypes": undefined,
              "render": [Function],
            },
          },
          "type": Object {
            "$$typeof": Symbol(react.forward_ref),
            "defaultProps": Object {
              "fluid": false,
            },
            "displayName": "Container",
            "render": [Function],
          },
        },
      ],
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
